[Feature] Bazaar Unique Item ID and Offline Trading

neckkola opened the pull request:
# Description

This update allows for proper unique item ids within the bazaar, utilizing the RoF2 model of a 17 character string.  It is generated using a random hash generator which is unique across zones.  It also allows for easier determination of multi-instance bazaars to ensure 100% accuracy in unique ids.

Further, it removes the workaround for hot bars.  Previously, there was code upon character login to ensure that item ids (which were simple uint32s, reused across zones) were not reused to ensure that hot bar items would not 'change'.  This code was expensive now that there are large bags.  This is removed.

On first boot of world, a conversion routine runs to create unique ids for all character inventory items.  This is done once, though I left the routine in zone inventory load as a fail safe.

Trader, Buyer, Barter, Parcels, Evolving Items have been updated to reflect the new unique item id.

This feature also allows off line trading for traders/buyers.  Using the 'offline mode' will remove the active player, and place them in offline mode.  /who and /who all have been updated along with the Guild Status window.  Therefore, there are login server changes to manage the offline status, displaying the appropriate messaging to clients when they login, should they have an offline character active.

I have also rewritten most of the bazaar code to be more efficient and better manage resources.  I have also added more graceful transaction rollbacks for when a bazaar transaction fails.

This is not ready for merge.  I would like to ask if there are any interested parties in testing?  I have been trough many, though would always appreciate another set of eyes, and attempts to break things. :-)  Testing with the alternate bazaar functionality (multi-instance) would also be appreciated.

It does not have 'Direct to Inventory' functionality (which the previous version also did not have) though THJ has implemented so I will see if they will allow me to add to this in a future version.

Thank you!

## Type of change

- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [X] This change requires a documentation update

# Testing
I have tested locally, and the offline functionality has been tested in a prod environment (not with the new unique item id) for several weeks.  Is ok at present.

Clients tested: 
RoF2

# Checklist

- [X] I have tested my changes
- [X] I have performed a self-review of my code. Ensuring variables, functions and methods are named in a human-readable way, comments are added only where naming of variables, functions and methods can't give enough context.
- [WILL DO AFTER TESTING] I have made corresponding changes to the documentation (if applicable, if not delete this line)
- [X] I own the changes of my code and take responsibility for the potential issues that occur
- [X] If my changes make database schema changes, I have tested the changes on a local database (attach image). Updated version.h CURRENT_BINARY_DATABASE_VERSION to the new version. (Delete this if not applicable)

Valorith commented:
You sir, are amazing!

fryguy503 commented:
I do not believe the offline trader portion will fully function with the current public login server however, it will require a private login server or public server revamp. Neckkola can chime in on that.

fryguy503 commented on zone/zonedb.cpp:
```diff
@@ -354,8 +357,8 @@ std::unique_ptr<EQ::ItemInstance> ZoneDatabase::LoadSingleTraderItem(uint32 char
+	inst->SetUniqueID(serial_number);
+	//FIX inst->SetMerchantSlot(serial_number);
 	inst->SetPrice(cost);
 
```
Line 386 will need changed:

From
`"`char_id` = '{}' AND `item_id` = {}",` 

To:
`"`character_id` = '{}' AND `item_id` = {}",`

neckkola replied:
Updated these.  Found a few.

fryguy503 commented:
client_packet.cpp:834 will need changed:

From:
```
TraderRepository::DeleteWhere(database, fmt::format("`char_id` = '{}'", CharacterID()));
```

To:
```
TraderRepository::DeleteWhere(database, fmt::format("`character_id` = '{}'", CharacterID()));
```

fryguy503 commented:
I know it doesn't solve your issue, but just general feedback, been running this PR for a while now and no major issues.

Akkadius commented on common/bazaar.cpp:
```diff
@@ -262,16 +262,26 @@ Bazaar::GetSearchResults(
-	std::vector<std::string>                      trader_items_ids{};
+	std::unordered_set<std::string>               trader_items_ids{};
+
+//	auto const trader_results = TraderRepository::GetBazaarTraderDetails(db, search_criteria_trader, search.max_results);
```
Can we remove this comment ?

Akkadius commented on common/bazaar.cpp:
```diff
@@ -290,43 +300,71 @@ Bazaar::GetSearchResults(
-	if (all_entries.size() > search.max_results) {
-		all_entries.resize(search.max_results);
-	}
+	// for (auto const& t:item_results) {
+	// 	BazaarSearchResultsFromDB_Struct r{};
+	// 	r.count                   = 1;
+	// 	r.trader_id               = t.trader.character_id;
+	// 	r.item_unique_id          = t.trader.item_unique_id;
+	// 	r.cost                    = t.trader.item_cost;
+	// 	r.slot_id                 = t.trader.slot_id;
+	// 	r.charges                 = t.trader.item_charges;
+	// 	r.stackable               = t.stackable;
+	// 	r.icon_id                 = t.icon;
+	// 	r.trader_zone_id          = t.trader.char_zone_id;
+	// 	r.trader_zone_instance_id = t.trader.char_zone_instance_id;
+	// 	r.trader_entity_id        = t.trader.char_entity_id;
+	// 	r.item_name               = fmt::format("{:.63}\0", t.name);
+	// 	r.trader_name             = fmt::format("{:.63}\0", t.trader_name);
+	// 	r.item_stat               = t.stats;
+	//
+	// 	if (RuleB(Bazaar, UseAlternateBazaarSearch)) {
+	// 		if (convert ||
+	// 			char_zone_id != Zones::BAZAAR ||
+	// 			(char_zone_id == Zones::BAZAAR && r.trader_zone_instance_id != char_zone_instance_id)
+	// 			) {
+	// 			r.trader_id = TraderRepository::TRADER_CONVERT_ID + r.trader_zone_instance_id;
+	// 			}
+	// 	}
+	//
+	// 	all_entries.push_back(r);
+	// }
+
+	// if (all_entries.size() > search.max_results) {
```
Remove if going away

neckkola replied:
Updated

Akkadius commented on common/patches/rof2.cpp:
```diff
@@ -533,7 +533,7 @@ namespace RoF2
 			}
 			default: {
-				LogTrading("(RoF2) Unhandled action <red>[{}]", action);
+				//LogTradingDetail("(RoF2) Unhandled action <red>[{}]", action);
```
Remove if not keeping

neckkola replied:
Updated

Akkadius commented on common/patches/rof2.cpp:
```diff
@@ -622,10 +622,10 @@ namespace RoF2
 				break;
 			}
 			default: {
-				LogTrading(
```
Remove if not keeping

neckkola replied:
Updated

Akkadius commented on common/patches/rof2.cpp:
```diff
@@ -6335,13 +6354,14 @@ namespace RoF2
 			}
 			default: {
-				LogTrading("(RoF2) Unhandled action <red>[{}]", action);
+				//LogTradingDetail("(RoF2) Unhandled action <red>[{}]", action);
```
Remove if not keeping

Akkadius replied:
Remove if not keeping

Akkadius commented on common/patches/rof2.cpp:
```diff
@@ -6446,7 +6466,21 @@ namespace RoF2
+			hdr.unknown000[16] = '\0';
+		}
+
+		// strn0cpy(
+		// 	hdr.unknown000,
+		// 	inst->GetSerialNumber2().empty() ? "0000000000000000" : inst->GetSerialNumber2().c_str(),
+		// 	sizeof(hdr.unknown000)
+		// );
+		//hdr.unknown000[16] = '\0';
```
Remove if not keeping

Akkadius commented on common/patches/uf.cpp:
```diff
@@ -3618,7 +3618,7 @@ namespace UF
 				IN(action);
 				memcpy(emu->player_name, eq->player_name, sizeof(emu->player_name));
-				IN(serial_number);
+				//FIXIN(serial_number);
```
Remove if not keeping

Akkadius commented on common/repositories/inventory_snapshots_repository.h:
```diff
@@ -68,6 +95,227 @@ class InventorySnapshotsRepository: public BaseInventorySnapshotsRepository {
+			del_time -= RuleI(Character, InvSnapshotHistoryD) * 86400;
+		}
+
+		DeleteWhere(db, fmt::format("`character_id` = '{}' AND `time_index` <= '{}'", character_id, del_time));
```
Numbers shouldn't be quoted in queries

neckkola replied:
Updated

Akkadius commented on common/repositories/inventory_snapshots_repository.h:
```diff
@@ -68,6 +95,227 @@ class InventorySnapshotsRepository: public BaseInventorySnapshotsRepository {
+	static void ListCharacterInvSnapshots(Database &db, uint32 character_id, std::list<std::pair<uint32, int>> &is_list)
+	{
+		const std::string &query = fmt::format(
+			"SELECT `time_index`, COUNT(*) FROM `inventory_snapshots` WHERE "
```
Numbers shouldn't be quoted in queries, it throws off the query planner

neckkola replied:
Updated

Akkadius commented on common/repositories/inventory_snapshots_repository.h:
```diff
@@ -68,6 +95,227 @@ class InventorySnapshotsRepository: public BaseInventorySnapshotsRepository {
+	{
+		const std::string &query = fmt::format(
+			"SELECT `slot_id`, `item_id` FROM `inventory_snapshots` "
+			"WHERE `character_id` = '{}' AND `time_index` = '{}' ORDER BY `slot_id`",
```
Numbers shouldn't be quoted in queries, it throws off the query planner

neckkola replied:
Updated

Akkadius commented on common/repositories/inventory_snapshots_repository.h:
```diff
@@ -68,6 +95,227 @@ class InventorySnapshotsRepository: public BaseInventorySnapshotsRepository {
+	{
+		const std::string &query = fmt::format(
+			"SELECT `slot_id`, `item_id` FROM `inventory` WHERE "
+			"`character_id` = '{0}' AND `slot_id` NOT IN ("
```
Numbers shouldn't be quoted in queries, it throws off the query planner

Akkadius commented on common/repositories/inventory_snapshots_repository.h:
```diff
@@ -68,6 +95,227 @@ class InventorySnapshotsRepository: public BaseInventorySnapshotsRepository {
+		uint32                          time_index = time(nullptr);
+		std::vector<InventorySnapshots> queue{};
+
+		auto inventory = InventoryRepository::GetWhere(db, fmt::format("`character_id` = '{}'", character_id));
```
Numbers shouldn't be quoted in queries, it throws off the query planner

neckkola replied:
Updated

Akkadius commented on common/repositories/inventory_snapshots_repository.h:
```diff
@@ -68,6 +95,227 @@ class InventorySnapshotsRepository: public BaseInventorySnapshotsRepository {
+	{
+		InventoryRepository::DeleteWhere(db, fmt::format("`character_id` = '{}'", character_id));
+
+		auto snapshot = GetWhere(db, fmt::format("`character_id` = '{}' AND `time_index` = '{}'", character_id, timestamp));
```
Numbers shouldn't be quoted in queries, it throws off the query planner

neckkola replied:
Updated

Akkadius commented on common/repositories/inventory_snapshots_repository.h:
```diff
@@ -68,6 +95,227 @@ class InventorySnapshotsRepository: public BaseInventorySnapshotsRepository {
+
+		std::vector<InventoryRepository::Inventory> queue{};
+		for (auto const &i: snapshot) {
+			auto inventory_entry                = InventoryRepository::NewEntity();
```
Just make this `e` vs `inventory_entry` - makes the code needlessly verbose when its a receiver of a complex type

Akkadius commented on common/repositories/trader_repository.h:
```diff
@@ -113,15 +118,15 @@ class TraderRepository : public BaseTraderRepository {
 				fmt::format(
-					"char_id = '{}' AND item_id = '{}'",
-					char_id,
+					"character_id = '{}' AND item_id = '{}'",
```
Numbers shouldn't be quoted in queries, it throws off the query planner

Akkadius replied:
Remove quotes from numbers in query please

Akkadius commented on common/repositories/trader_repository.h:
```diff
@@ -155,8 +160,8 @@ class TraderRepository : public BaseTraderRepository {
-			"SELECT t.char_id, t.item_id, t.serialnumber, t.charges, t.item_cost, t.slot_id, t.entity_id FROM trader AS t "
-			"WHERE t.entity_id = {} AND t.item_id = {} AND t.item_cost = {} "
+			"SELECT t.character_id, t.item_id, t.item_unique.id, t.charges, t.item_cost, t.slot_id, t.entity_id FROM trader AS t "
+			"WHERE t.entity_id = '{}' AND t.item_id = '{}' AND t.item_cost = '{}' "
```
Numbers shouldn't be quoted in queries, it throws off the query planner

Akkadius commented on common/repositories/trader_repository.h:
```diff
@@ -168,41 +173,69 @@ class TraderRepository : public BaseTraderRepository {
 		const auto trader_item = GetWhere(
 			db,
-			fmt::format("char_id = '{}' AND item_sn = '{}' ", char_id, serial_number)
+			fmt::format("`item_unique_id` = '{}' ", item_unique_id)
```
Numbers shouldn't be quoted in queries, it throws off the query planner

Akkadius commented on common/repositories/trader_repository.h:
```diff
@@ -168,41 +173,69 @@ class TraderRepository : public BaseTraderRepository {
+		std::vector<Trader> all_entries{};
+
+		const auto query = fmt::format(
+			"UPDATE trader t1 SET t1.`item_cost` = '{}', t1.`listing_date` = FROM_UNIXTIME({}) WHERE t1.`item_id` = "
```
Numbers shouldn't be quoted in queries, it throws off the query planner

Akkadius replied:
Remove quotes from numbers in query please

Akkadius commented on common/repositories/trader_repository.h:
```diff
@@ -168,41 +173,69 @@ class TraderRepository : public BaseTraderRepository {
+		all_entries = GetWhere(
+			db,
+			fmt::format(
+				"`item_id` = (SELECT t1.`item_id` FROM trader t1 WHERE t1.`item_unique_id` = '{}');",
```
Numbers shouldn't be quoted in queries, it throws off the query planner

Akkadius commented on common/repositories/trader_repository.h:
```diff
@@ -289,14 +315,45 @@ class TraderRepository : public BaseTraderRepository {
 	{
 		std::vector<BazaarTraderSearch_Struct> all_entries{};
 
+		// auto query_2 = fmt::format(
+		// "WITH ranked_trader_items AS ("
+		// 	"SELECT trader.id, trader.character_id, trader.item_id, trader.item_unique_id, trader.augment_one, "
+		// 	"trader.augment_two, trader.augment_three, trader.augment_four, trader.augment_five, trader.augment_six, "
+		// 	"trader.item_charges, trader.item_cost, trader.slot_id, trader.char_entity_id, trader.char_zone_id, "
+		// 	"trader.char_zone_instance_id, trader.active_transaction, c.`name`, "
+		// 	"items.name AS n1, items.stackable, items.icon, {}, "
+		// 	"ROW_NUMBER() OVER (PARTITION BY trader.character_id) AS row_num "
+		// 	"FROM trader "
+		// 	"INNER JOIN character_data AS c ON trader.character_id = c.id "
+		// 	"JOIN peq642024_content.items AS items ON trader.item_id = items.id "
+		// 	"WHERE items.`name` LIKE '%{}%' AND {} AND {}"
+		// ") "
+		// "SELECT * FROM ranked_trader_items "
+		// "WHERE row_num <= '{}';",
+		// field_criteria_items,
+		// Strings::Escape(name),
+		// where_criteria_items,
+		// search_criteria_trader,
+		// max_results
+		// );
```
Remove if not keeping

Akkadius commented on common/repositories/trader_repository.h:
```diff
@@ -328,12 +385,59 @@ class TraderRepository : public BaseTraderRepository {
 			e.trader.char_zone_instance_id = row[15] ? static_cast<int32_t>(atoi(row[15])) : 0;
 			e.trader.active_transaction    = row[16] ? static_cast<uint8_t>(strtoul(row[16], nullptr, 10)) : 0;
 			e.trader_name                  = row[17] ? row[17] : std::string("");
+			// e.name                         = row[18] ? row[18] : "";
+			// e.stackable                    = atoi(row[19]) ? true : false;
+			// e.icon                         = row[20] ? static_cast<int32_t>(atoi(row[20])) : 0;
+			// e.stats                        = row[21] ? static_cast<int32_t>(atoi(row[21])) : 0;
```
Remove if not keeping, not sure why we're manually modifying repositories

Akkadius commented on common/repositories/trader_repository.h:
```diff
@@ -328,12 +385,59 @@ class TraderRepository : public BaseTraderRepository {
+		auto trader_query = fmt::format(
+			"SELECT t.id, t.character_id, t.char_zone_id, t.char_zone_instance_id "
+			"FROM trader AS t "
+			"WHERE t.character_id IN(SELECT c.id FROM character_data AS c WHERE c.account_id = '{}') "
```
Numbers shouldn't be quoted in queries, it throws off the query planner

Akkadius commented on common/repositories/trader_repository.h:
```diff
@@ -328,12 +385,59 @@ class TraderRepository : public BaseTraderRepository {
+		auto buyer_query = fmt::format(
+			"SELECT t.id, t.char_id, t.char_zone_id, t.char_zone_instance_id "
+			"FROM buyer AS t "
+			"WHERE t.char_id IN(SELECT c.id FROM character_data AS c WHERE c.account_id = '{}') "
```
Numbers shouldn't be quoted in queries, it throws off the query planner

Akkadius commented on common/inventory_profile.cpp:
```diff
@@ -252,11 +252,18 @@ int16 EQ::InventoryProfile::PutItem(int16 slot_id, const ItemInstance& inst)
 		return slot_id;
 	}
 
+	// if (inst.GetSerialNumber2().empty()) {
+	// 	inst.GenerateUniqueSerialNumber();
+	// }
```
Remove if not keeping

Akkadius commented on common/patches/rof2.cpp:
```diff
@@ -622,10 +622,10 @@ namespace RoF2
-					"(RoF2) Unhandled action <red>[{}]",
-					in->action
-				);
+				// LogTradingDetail(
+				// 	"(RoF2) Unhandled action <red>[{}]",
+				// 	in->action
+				// );
```
Remove if not keeping

Akkadius commented on common/repositories/inventory_snapshots_repository.h:
```diff
@@ -46,16 +47,42 @@ class InventorySnapshotsRepository: public BaseInventorySnapshotsRepository {
+		}
+
+		if (count < 0) {
+			return -3;
```
Magic number ?

neckkola replied:
It would appear so.  I moved the zonedb functions into the repo only, and did not change the detail.

Akkadius commented on common/repositories/inventory_snapshots_repository.h:
```diff
@@ -46,16 +47,42 @@ class InventorySnapshotsRepository: public BaseInventorySnapshotsRepository {
+	{
+		const std::string &query = fmt::format(
+			"SELECT COUNT(*) FROM (SELECT * FROM `inventory_snapshots` a WHERE "
+			"`character_id` = '{}' GROUP BY `time_index`) b",
```
Remove quotes from numbers in query please

Akkadius commented on common/repositories/inventory_snapshots_repository.h:
```diff
@@ -68,6 +95,227 @@ class InventorySnapshotsRepository: public BaseInventorySnapshotsRepository {
+	{
+		const std::string &query = fmt::format(
+			"SELECT `time_index`, COUNT(*) FROM `inventory_snapshots` WHERE "
+			"`character_id` = '{}' GROUP BY `time_index` ORDER BY `time_index` DESC",
```
Remove quotes from numbers in query please

Akkadius commented on common/repositories/inventory_snapshots_repository.h:
```diff
@@ -68,6 +95,227 @@ class InventorySnapshotsRepository: public BaseInventorySnapshotsRepository {
+		}
+
+		const std::string &query = fmt::format(
+			"SELECT * FROM `inventory_snapshots` WHERE `character_id` = '{}' "
```
Remove quotes from numbers in query please

neckkola replied:
Updated

Akkadius commented on common/repositories/inventory_snapshots_repository.h:
```diff
@@ -68,6 +95,227 @@ class InventorySnapshotsRepository: public BaseInventorySnapshotsRepository {
+	{
+		const std::string &query = fmt::format(
+			"SELECT `slot_id`, `item_id` FROM `inventory_snapshots` "
+			"WHERE `character_id` = '{}' AND `time_index` = '{}' ORDER BY `slot_id`",
```
Remove quotes from numbers in query please

neckkola replied:
Updated

Akkadius commented on common/repositories/inventory_snapshots_repository.h:
```diff
@@ -68,6 +95,227 @@ class InventorySnapshotsRepository: public BaseInventorySnapshotsRepository {
+	{
+		const std::string &query = fmt::format(
+			"SELECT slot_id, item_id FROM `inventory_snapshots` "
+			"WHERE `time_index` = '{0}' AND `character_id` = '{1}' AND `slot_id` NOT IN ("
```
Remove quotes from numbers in query please

Akkadius commented on common/repositories/inventory_snapshots_repository.h:
```diff
@@ -68,6 +95,227 @@ class InventorySnapshotsRepository: public BaseInventorySnapshotsRepository {
+			"SELECT slot_id, item_id FROM `inventory_snapshots` "
+			"WHERE `time_index` = '{0}' AND `character_id` = '{1}' AND `slot_id` NOT IN ("
+			"SELECT a.`slot_id` FROM `inventory_snapshots` a JOIN `inventory` b USING (`slot_id`, `item_id`) "
+			"WHERE a.`time_index` = '{0}' AND a.`character_id` = '{1}' AND b.`character_id` = '{1}')",
```
Remove quotes from numbers in query please

neckkola replied:
Updated

Akkadius commented on common/repositories/inventory_snapshots_repository.h:
```diff
@@ -68,6 +95,227 @@ class InventorySnapshotsRepository: public BaseInventorySnapshotsRepository {
+			"SELECT `slot_id`, `item_id` FROM `inventory` WHERE "
+			"`character_id` = '{0}' AND `slot_id` NOT IN ("
+			"SELECT a.`slot_id` FROM `inventory` a JOIN `inventory_snapshots` b USING (`slot_id`, `item_id`) "
+			"WHERE b.`time_index` = '{1}' AND b.`character_id` = '{0}' AND a.`character_id` = '{0}')",
```
Remove quotes from numbers in query please

Akkadius commented on common/repositories/trader_repository.h:
```diff
@@ -130,8 +135,8 @@ class TraderRepository : public BaseTraderRepository {
 				fmt::format(
-					"char_id = '{}' AND item_id = '{}' AND item_charges = '{}'",
-					char_id,
+					"character_id = '{}' AND item_id = '{}' AND item_charges = '{}'",
```
Remove quotes from numbers in query please

Akkadius commented on common/repositories/trader_repository.h:
```diff
@@ -155,8 +160,8 @@ class TraderRepository : public BaseTraderRepository {
 		auto query   = fmt::format(
-			"SELECT t.char_id, t.item_id, t.serialnumber, t.charges, t.item_cost, t.slot_id, t.entity_id FROM trader AS t "
-			"WHERE t.entity_id = {} AND t.item_id = {} AND t.item_cost = {} "
+			"SELECT t.character_id, t.item_id, t.item_unique.id, t.charges, t.item_cost, t.slot_id, t.entity_id FROM trader AS t "
```
Remove quotes from numbers in query please

Akkadius commented on common/item_instance.h:
```diff
@@ -228,8 +243,13 @@ namespace EQ
+		void               SetSerialNumber(int32 id) { m_SerialNumber = id; }
+		const std::string &GetSerialNumber2() const { return m_unique_id; }
+		const std::string &GetUniqueID() const { return m_unique_id; }
+		//std::string        &GetSerialNumber2() const { return m_serial_number2; }
```
Remove if not keeping

Akkadius commented on loginserver/login_util/login_opcodes_sod.conf:
```diff
@@ -11,3 +11,5 @@ OP_Poll=0x0029
 OP_EnterChat=0x000f
 OP_PollResponse=0x0011
+OP_CancelOfflineTrader=0x0016
+ OP_CancelOfflineTraderResponse=0x0030
```
Appears to be an extra space here

Akkadius commented on world/clientlist.cpp:
```diff
@@ -742,7 +755,14 @@ void ClientList::SendWhoAll(uint32 fromid,const char* to, int16 admin, Who_All_S
+					rankstring = 12315;
+				}
+				else if (cle->GetBuyer()) {
+					rankstring = 6056;
```
Magic number? Can we constant this off somewhere

Akkadius commented on world/clientlist.cpp:
```diff
@@ -742,7 +755,14 @@ void ClientList::SendWhoAll(uint32 fromid,const char* to, int16 admin, Who_All_S
-				} else if (cle->GetGM()) {
+				}
+				else if (cle->GetTrader()) {
+					rankstring = 12315;
```
Magic number? Can we constant this off somewhere

Akkadius commented on zone/client.h:
```diff
@@ -350,7 +351,7 @@ class Client : public Mob
 	void SendBuyerPacket(Client* Buyer);
 	void SendBuyerToBarterWindow(Client* buyer, uint32 action);
-	GetItems_Struct* GetTraderItems();
+	GetItems2_Struct* GetTraderItems();
```
Was this temporary? I'd prefer to not use GetThing2 GetThing3 and just have GetThing

Akkadius commented on zone/client.h:
```diff
@@ -315,7 +315,8 @@ class Client : public Mob
 
 	void TraderEndTrader();
-	void TraderPriceUpdate(const EQApplicationPacket *app);
+	//void TraderPriceUpdate(const EQApplicationPacket *app);
```
Remove if not keeping

Akkadius commented on zone/client.h:
```diff
@@ -2429,6 +2450,70 @@ class Client : public Mob
+ 		return Mob::GetMob();
+ 	}
+
+ 	void Clone(Client& in)
```
What is this for ?

Akkadius replied:
I see it in another part of the PR, for the offline functionality

neckkola replied:
It is a helper to clone a client

Akkadius commented on zone/client_packet.cpp:
```diff
@@ -853,6 +855,54 @@ void Client::CompleteConnect()
 	}
 
+	auto offline_transactions_trader = CharacterOfflineTransactionsRepository::GetWhere(
+		database, fmt::format("`character_id` = '{}' AND `type` = '{}'", CharacterID(), TRADER_TRANSACTION)
```
Please remove quotes from numbers in query

Akkadius commented on zone/client_packet.cpp:
```diff
@@ -853,6 +855,54 @@ void Client::CompleteConnect()
+	}
+
+	auto offline_transactions_buyer = CharacterOfflineTransactionsRepository::GetWhere(
+		database, fmt::format("`character_id` = '{}' AND `type` = '{}'", CharacterID(), BUYER_TRANSACTION)
```
Please remove quotes from numbers in query

Akkadius commented on zone/client_packet.cpp:
```diff
@@ -15443,7 +15493,7 @@ void Client::Handle_OP_Trader(const EQApplicationPacket *app)
 		}
 		default: {
-			LogError("Unknown size for OP_Trader: [{}]\n", app->size);
+			//LogTradingDetail("Unknown size for OP_Trader: [{}]", app->size);
```
Remove if we're not going to keep this

Akkadius commented on zone/trading.cpp:
```diff
@@ -1075,25 +1060,100 @@ EQ::ItemInstance *Client::FindTraderItemBySerialNumber(int32 SerialNumber)
+	return nullptr;
+}
+
+std::vector<EQ::ItemInstance *> Client::FindTraderItemsByUniqueID(const char* unique_id)
```
Any reason we don't just use std::string for the character type ?

Akkadius commented on zone/trading.cpp:
```diff
@@ -1075,25 +1060,100 @@ EQ::ItemInstance *Client::FindTraderItemBySerialNumber(int32 SerialNumber)
+	return nullptr;
+}
+
+EQ::ItemInstance *Client::FindTraderItemByUniqueID(const char* unique_id)
```
Any reason we don't just use std::string for the character type ?

Akkadius commented on zone/trading.cpp:
```diff
@@ -1075,25 +1060,100 @@ EQ::ItemInstance *Client::FindTraderItemBySerialNumber(int32 SerialNumber)
+				// we already have the parent bag and a contents iterator..why not just iterate the bag!??
+				slot_id = EQ::InventoryProfile::CalcSlotId(i, x);
+				item    = GetInv().GetItem(slot_id);
+				if (item) {
+					if (item->GetUniqueID().compare(unique_id) == 0) {
```
Could these two if blocks be combined ?

```if (item && item->GetUniqueID().compare(unique_id) == 0)```

Akkadius commented on zone/trading.cpp:
```diff
@@ -1075,25 +1060,100 @@ EQ::ItemInstance *Client::FindTraderItemBySerialNumber(int32 SerialNumber)
+				// we already have the parent bag and a contents iterator..why not just iterate the bag!??
+				slot_id = EQ::InventoryProfile::CalcSlotId(i, x);
+				item    = GetInv().GetItem(slot_id);
+				if (item) {
+					if (item->GetUniqueID().compare(unique_id) == 0) {
```
Could these two if blocks be combined ?

```if (item && item->GetUniqueID().compare(unique_id) == 0)```

Akkadius commented on zone/trading.cpp:
```diff
@@ -1075,25 +1060,100 @@ EQ::ItemInstance *Client::FindTraderItemBySerialNumber(int32 SerialNumber)
 	const EQ::ItemInstance *item   = nullptr;
 	int16                  slot_id = INVALID_INDEX;
-	auto                   gis     = new GetItems_Struct{0};
+	auto                   gis     = new GetItems2_Struct{0};
```
Can we make this just "GetItems"

Akkadius commented on zone/trading.cpp:
```diff
@@ -1238,7 +1300,7 @@ void Client::FindAndNukeTraderItem(int32 serial_number, int16 quantity, Client *
 		if (charges <= quantity || (charges <= 0 && quantity == 1) || !stackable) {
 			DeleteItemInInventory(slot_id, quantity);
-			auto   trader_items = TraderRepository::GetWhere(database, fmt::format("`char_id` = '{}'", CharacterID()));
+			auto   trader_items = TraderRepository::GetWhere(database, fmt::format("`character_id` = '{}'", CharacterID()));
```
Please remove the quote around the number param in the query

Akkadius commented on zone/trading.cpp:
```diff
@@ -1247,15 +1309,15 @@ void Client::FindAndNukeTraderItem(int32 serial_number, int16 quantity, Client *
 			std::vector<TraderRepository::Trader> delete_queue{};
 			for (int i = 0; i < item_limit; i++) {
-				if (test_slot && trader_items.at(i).item_sn == serial_number) {
+				if (test_slot && trader_items.at(i).item_unique_id.compare(item_unique_id) == 0) {
```
trader_items.at(i) && trader_items.at(i).item_unique_id.compare(item_unique_id) == 0

Akkadius commented on zone/trading.cpp:
```diff
@@ -1698,12 +1766,12 @@ static void UpdateTraderCustomerPriceChanged(
-					tdis->item_id = trader_items.at(i).item_sn;
-				}
-				tdis->item_id = trader_items.at(i).item_sn;
+				//FIX else {
+				// 	tdis->item_id = trader_items.at(i).item_sn;
+				// }
+				//tdis->item_id = trader_items.at(i).item_sn;
```
Remove if we're not keeping this

Akkadius commented on zone/zonedb.cpp:
```diff
@@ -354,8 +357,8 @@ std::unique_ptr<EQ::ItemInstance> ZoneDatabase::LoadSingleTraderItem(uint32 char
-	inst->SetSerialNumber(serial_number);
-	inst->SetMerchantSlot(serial_number);
+	inst->SetUniqueID(unique_item_id);
+	//FIX inst->SetMerchantSlot(serial_number);
```
remove if we don't plan on keeping this

Akkadius commented on zone/inventory.cpp:
```diff
@@ -650,8 +650,8 @@ bool Client::SummonItem(uint32 item_id, int16 charges, uint32 aug1, uint32 aug2,
-		PushItemOnCursor(*inst);
-		SendItemPacket(EQ::invslot::slotCursor, inst, ItemPacketLimbo);
+		PushItemOnCursor(*inst, true);
+		//SendItemPacket(EQ::invslot::slotCursor, inst, ItemPacketLimbo);
```
Remove if we're not going to keep. Although we was this removed ? Is the true to PushItemOnCursor making up for this ?

neckkola replied:
yes, it does.

Akkadius commented on common/patches/rof2.cpp:
```diff
@@ -6335,13 +6354,14 @@ namespace RoF2
 				IN_str(seller_name);
 				IN_str(item_name);
-				IN_str(serial_number);
+				//IN_str(serial_number);
```
Remove comment ?

Akkadius commented on common/patches/uf.cpp:
```diff
@@ -338,7 +338,7 @@ namespace UF
 					VARSTRUCT_ENCODE_TYPE(uint32, bufptr, 1);
 					VARSTRUCT_ENCODE_TYPE(int32, bufptr, row->item_id);
-					VARSTRUCT_ENCODE_TYPE(int32, bufptr, row->serial_number);
+					//FIX VARSTRUCT_ENCODE_TYPE(int32, bufptr, row->serial_number);
```
Remove if not keeping

Akkadius commented on common/patches/titanium.cpp:
```diff
@@ -228,7 +228,7 @@ namespace Titanium
 					bufptr += 4;
 					VARSTRUCT_ENCODE_TYPE(int32, bufptr, row->item_id);
-					VARSTRUCT_ENCODE_TYPE(int32, bufptr, row->serial_number);
+					//FIX VARSTRUCT_ENCODE_TYPE(int32, bufptr, row->serial_number);
```
Remove

neckkola commented:
Good evening.  I have updated based on comments and rebased to latest master.  I will need to do some testing next.

Akkadius commented:
> Good evening. I have updated based on comments and rebased to latest master. I will need to do some testing next.

Thank you Neck, have you had a chance to test? Do you want me to do another pass or wait?

neckkola commented on common/patches/rof2.cpp:
```diff
@@ -439,7 +439,7 @@ namespace RoF2
 			}
 			default: {
-				LogTradingDetail("Unhandled action <red>[{}]", sub_action);
+				//LogTradingDetail("Unhandled action <red>[{}]", sub_action);
```
Updated

neckkola commented on common/patches/rof2.cpp:
```diff
@@ -6189,18 +6220,15 @@ namespace RoF2
 			}
 			default: {
-				LogTrading("(RoF2) Unhandled action <red>[{}]", action);
+				//LogTradingDetail("(RoF2) Unhandled action <red>[{}]", action);
```
Updated

neckkola commented on common/patches/titanium.cpp:
```diff
@@ -2529,7 +2529,7 @@ namespace Titanium
 				IN(action);
 				memcpy(emu->player_name, eq->player_name, sizeof(emu->player_name));
-				IN(serial_number);
+				//FIXIN(serial_number);
```
Updated

neckkola commented on common/repositories/inventory_snapshots_repository.h:
```diff
@@ -68,6 +95,227 @@ class InventorySnapshotsRepository: public BaseInventorySnapshotsRepository {
+
+	static bool RestoreCharacterInvSnapshot(Database &db, uint32 character_id, uint32 timestamp)
+	{
+		InventoryRepository::DeleteWhere(db, fmt::format("`character_id` = '{}'", character_id));
```
Updated

neckkola commented on common/item_instance.h:
```diff
@@ -346,34 +366,77 @@ namespace EQ
+		}
+
+	public:
+		static std::string generate()
```
Yes, that is the intent.

neckkola commented on common/repositories/account_repository.h:
```diff
@@ -107,6 +107,45 @@ class AccountRepository: public BaseAccountRepository {
+
+	static void ClearAllOfflineStatus(Database& db)
+	{
+		auto query = fmt::format("UPDATE {} SET `offline` = '0' WHERE `offline` = '1';",
```
Updated

neckkola commented on common/repositories/account_repository.h:
```diff
@@ -107,6 +107,45 @@ class AccountRepository: public BaseAccountRepository {
+		auto query = fmt::format("SELECT a.`offline` "
+			"FROM `account` AS a "
+			"INNER JOIN character_data AS c ON c.account_id = a.id "
+			"WHERE c.id = '{}'",
```
Updated

neckkola commented on common/repositories/inventory_snapshots_repository.h:
```diff
@@ -46,16 +47,42 @@ class InventorySnapshotsRepository: public BaseInventorySnapshotsRepository {
+		const int64 count = Strings::ToBigInt(row[0]);
+
+		if (count > std::numeric_limits<int>::max()) {
+			return -2;
```
It would appear so.  I moved the zonedb functions into the repo only, and did not change the detail.

neckkola commented on common/repositories/inventory_snapshots_repository.h:
```diff
@@ -68,6 +95,227 @@ class InventorySnapshotsRepository: public BaseInventorySnapshotsRepository {
+			del_time -= RuleI(Character, InvSnapshotHistoryD) * 86400;
+		}
+
+		DeleteWhere(db, fmt::format("`character_id` = '{}' AND `time_index` <= '{}'", character_id, del_time));
```
Updated

neckkola commented on common/repositories/inventory_snapshots_repository.h:
```diff
@@ -68,6 +95,227 @@ class InventorySnapshotsRepository: public BaseInventorySnapshotsRepository {
+	{
+		const std::string &query = fmt::format(
+			"SELECT `slot_id`, `item_id` FROM `inventory` WHERE "
+			"`character_id` = '{0}' AND `slot_id` NOT IN ("
```
Updated

neckkola commented on common/repositories/inventory_snapshots_repository.h:
```diff
@@ -68,6 +95,227 @@ class InventorySnapshotsRepository: public BaseInventorySnapshotsRepository {
+		uint32                          time_index = time(nullptr);
+		std::vector<InventorySnapshots> queue{};
+
+		auto inventory = InventoryRepository::GetWhere(db, fmt::format("`character_id` = '{}'", character_id));
```
Updated

neckkola commented on common/repositories/inventory_snapshots_repository.h:
```diff
@@ -68,6 +95,227 @@ class InventorySnapshotsRepository: public BaseInventorySnapshotsRepository {
+
+		std::vector<InventoryRepository::Inventory> queue{};
+		for (auto const &i: snapshot) {
+			auto inventory_entry                = InventoryRepository::NewEntity();
```
Updated

neckkola commented:
I have rebased, and updated based on comments.  Ran some basic tests.  I do need to update the GetSerialNumber functions to remove them and clean up that side.  Will work on that later.

neckkola commented:
Ran more tests.  Corrected a few issues and also implemented shared bank for unique ids.  Should be ready for another review. Thank you.

Kinglykrab reviewed (approved):
This looks great, will be an extremely welcome addition.

neckkola commented:
May need some assistance with the pipeline.  Seems like something is amiss.  Builds locally. Appreciate if someone can take a look.
